### PR TITLE
[M4] key pair validation

### LINKED ISSUE #102: M4 - Integrate Private key data structure and validate keypairs from wallet
Desc:
As a prerequisite of transaction sending, keypairs needs to be imported with validation

Tasks:

- Validate keypairs

### PR DESCRIPTION
**Pull request by hanabi1224**
**Summary of changes**
Changes introduced in this pull request:
- add data structure for private key
- key pair validation functionality

**Reference issue to close (if applicable)**

Closes #102 

**Other information and links**

### CODE REVIEW

File: crypto/Cargo.toml [hunk 1]
```diff
@@ -10,8 +10,13 @@ license = "Apache-2.0"
 bin-prot = {path = "../protocol/bin-prot"}
 wire-type = {path = "../protocol/wire-type"}
 
+mina-curves = {git = "https://github.com/o1-labs/proof-systems", rev = "6af257b793d80f900675acc01b16fd4bf6375268"}
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply crypto/Cargo.toml:13]: That's where the mina 'pasta' elliptic curve is defined

File: crypto/src/signature.rs [hunk 2]
```diff
@@ -25,6 +62,30 @@ impl Base58Encodable for PublicKey {
...
+    fn from(p: pallas::Affine) -> Self {
+        let x = p.x;
+        let y = p.y;
+        let x_big: BigInteger256 = x.into();
+        let y_big: BigInteger256 = y.into();
+        let x_bytes_vec = x_big.to_bytes_le();
+        let mut x_bytes = [0; 32];
+        x_bytes.copy_from_slice(x_bytes_vec.as_slice());
+        Self {
+            poly: CompressedCurvePoint {
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply crypto/src/signature.rs]: That's mina representation of a point on mina 'pasta' affine curve

File: crypto/src/signature.rs [hunk 3]
```diff
@@ -1,11 +1,48 @@
...
+
+    pub fn validate(&self, public_key: &PublicKey) -> Result<bool, Error> {
+        Ok(self.derive_public_key()?.borrow() == public_key)
+    }
+
+    pub fn try_to_bigint(&self) -> Result<BigInteger256, Error> {
+        BigInteger256::read(self.0.as_slice()).map_err(Error::IoError)
+    }
+
+    pub fn try_to_fq(&self) -> Result<Fq, Error> {
```
[the remark this answers is not in the document]
Comment (hanabi1224) [reply crypto/src/signature.rs:40]: That's the scalar file type on mina 'pasta' elliptic curve,  checkout https://github.com/o1-labs/proof-systems/tree/master/curves/src/pasta/fields

**Comment by jspada** [comment]
Hey all.  Unfortunately this code is a duplication of effort.  As mentioned in the meetings, Mina's crypto team has written a signing library for mina-rs to use without getting into all the technical details of the signing algorithm, keypairs, secret keys, etc.  Additionally anything related to verification should not be worked on now as this layer is in a state of flux and assigned to another milestone.

For now please stub anything related to verification/keypairs/etc.  When you start the verification milestone you can create a wrapper around Mina's official rust crypto library in order to avoid re-implementing this code.  It is highly important to use our version to assure correct use of cryptography.

**Comment by hanabi1224** [comment]
Thank you @jspada , will discard this one

**Pull request closed without merging**